[inductor] Rewrite Triton templates + epilogue fusion (retry)

jansel opened the pull request:
This reverts commit 94262efc7d381ace82aa74ed2f5f5ec76f8fca95 to reland #91105 / #90738.

Fixes https://github.com/pytorch/torchdynamo/issues/2015



cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

ngimel commented:
The previous version regressed perf of hf models that don't use triton matmul templates, because it broke eager linear + bias fusion. I had a hot fix for that 
```
diff --git a/torch/_inductor/kernel/mm.py b/torch/_inductor/kernel/mm.py
index 431f8c13079..8bc3745cd04 100644
--- a/torch/_inductor/kernel/mm.py
+++ b/torch/_inductor/kernel/mm.py
@@ -95,15 +95,20 @@ def tuned_mm(mat1, mat2, *, layout=None):
 
 @register_lowering(aten.addmm)
 def tuned_addmm(inp, mat1, mat2, *, alpha=1, beta=1, layout=None):
-    m, n, k, layout, mat1, mat2, inp = mm_args(mat1, mat2, inp, layout=layout)
+    m, n, k, layout, mat1, mat2, inp_expanded = mm_args(mat1, mat2, inp, layout=layout)
+    # don't expand inp to make sure fused addmm from cublasLt is used
+    if not use_triton_template(layout):
+        choices = [aten_addmm.bind((inp, mat1, mat2), layout, alpha=alpha, beta=beta)]
+        return autotune_select_algorithm(choices, [inp, mat1, mat2], layout)
 
+    # TODO this is not quite fair benchmarking because we won't use fused cublasLt addmm
     # options to tune from
-    choices = [aten_addmm.bind((inp, mat1, mat2), layout, alpha=alpha, beta=beta)]
+    choices = [aten_addmm.bind((inp_expanded, mat1, mat2), layout, alpha=alpha, beta=beta)]
     if use_triton_template(layout):
         for config in mm_configs():
             choices.append(
                 mm_template.generate(
-                    (inp, mat1, mat2),
+                    (inp_expanded, mat1, mat2),
                     layout,
                     **mm_options(config, k, layout),
                     prefix_args=1,
@@ -111,4 +116,4 @@ def tuned_addmm(inp, mat1, mat2, *, alpha=1, beta=1, layout=None):
                 )
             )
 
-    return autotune_select_algorithm(choices, [inp, mat1, mat2], layout)
+    return autotune_select_algorithm(choices, [inp_expanded, mat1, mat2], layout)
```
can you please put it in here?

jansel committed:
[inductor] Rewrite Triton templates + epilogue fusion (retry) (#91575)

Summary:
This reverts commit 94262efc7d381ace82aa74ed2f5f5ec76f8fca95 to reland https://github.com/pytorch/pytorch/issues/91105 / https://github.com/pytorch/pytorch/issues/90738.

Fixes https://github.com/pytorch/torchdynamo/issues/2015

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Pull Request resolved: https://github.com/pytorch/pytorch/pull/91575

Reviewed By: ngimel

Differential Revision: D42304332

Pulled By: jansel

fbshipit-source-id: 1eefc7320da5de7544d048c5b7ea8716930f31cf
